File: 04_BehaviouralDesignPatterns/05_StatePattern/cpp/Main.cpp
#include <iostream>
#include <memory>
#include <string>

// Forward declaration
class Order;

// State interface
class OrderState {
public:
    virtual void nextState(Order& order) = 0;
    virtual void cancelOrder(Order& order) = 0;
    virtual std::string getStateName() const = 0;
    virtual ~OrderState() = default;
};

// Order class (Context)
class Order {
private:
    std::unique_ptr<OrderState> currentState;

public:
    Order();
    void setState(std::unique_ptr<OrderState> state);
    void nextState();
    void cancelOrder();
    std::string getState() const;
};

// Concrete states
class OrderPlacedState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "ORDER_PLACED";
    }
};

class PreparingState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "PREPARING";
    }
};

class ReadyForDeliveryState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "READY_FOR_DELIVERY";
    }
};

class OutForDeliveryState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "OUT_FOR_DELIVERY";
    }
};

class DeliveredState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "DELIVERED";
    }
};

class CancelledState : public OrderState {
public:
    void nextState(Order& order) override;
    void cancelOrder(Order& order) override;
    std::string getStateName() const override {
        return "CANCELLED";
    }
};

// Order implementation
Order::Order() {
    currentState = std::make_unique<OrderPlacedState>();
}

void Order::setState(std::unique_ptr<OrderState> state) {
    currentState = std::move(state);
}

void Order::nextState() {
    currentState->nextState(*this);
}

void Order::cancelOrder() {
    currentState->cancelOrder(*this);
}

std::string Order::getState() const {
    return currentState->getStateName();
}

// State implementations
void OrderPlacedState::nextState(Order& order) {
    std::cout << "Order is now being prepared" << std::endl;
    order.setState(std::make_unique<PreparingState>());
}

void OrderPlacedState::cancelOrder(Order& order) {
    std::cout << "Cancelling order" << std::endl;
    order.setState(std::make_unique<CancelledState>());
}

void PreparingState::nextState(Order& order) {
    std::cout << "Order is now ready for delivery" << std::endl;
    order.setState(std::make_unique<ReadyForDeliveryState>());
}

void PreparingState::cancelOrder(Order& order) {
    std::cout << "Cancelling order" << std::endl;
    order.setState(std::make_unique<CancelledState>());
}

void ReadyForDeliveryState::nextState(Order& order) {
    std::cout << "Order is out for delivery" << std::endl;
    order.setState(std::make_unique<OutForDeliveryState>());
}

void ReadyForDeliveryState::cancelOrder(Order& order) {
    std::cout << "Cannot cancel order in ready for delivery state" << std::endl;
}

void OutForDeliveryState::nextState(Order& order) {
    std::cout << "Order has been delivered" << std::endl;
    order.setState(std::make_unique<DeliveredState>());
}

void OutForDeliveryState::cancelOrder(Order& order) {
    std::cout << "Cannot cancel order in out for delivery state" << std::endl;
}

void DeliveredState::nextState(Order& order) {
    std::cout << "No further state transitions available from delivered state" << std::endl;
}

void DeliveredState::cancelOrder(Order& order) {
    std::cout << "Cannot cancel delivered order" << std::endl;
}

void CancelledState::nextState(Order& order) {
    std::cout << "Cannot transition from cancelled state" << std::endl;
}

void CancelledState::cancelOrder(Order& order) {
    std::cout << "Order is already cancelled" << std::endl;
}

int main() {
    Order order;
    std::cout << "Initial state: " << order.getState() << std::endl;
    
    order.nextState();  // ORDER_PLACED -> PREPARING
    order.nextState();  // PREPARING -> READY_FOR_DELIVERY
    order.cancelOrder(); // Cannot cancel in READY_FOR_DELIVERY
    order.nextState();  // READY_FOR_DELIVERY -> OUT_FOR_DELIVERY
    order.nextState();  // OUT_FOR_DELIVERY -> DELIVERED
    order.nextState();  // No further transitions
    order.cancelOrder(); // Cannot cancel delivered order

    Order order2;
    std::cout << "\nNew order - Initial state: " << order2.getState() << std::endl;
    order2.cancelOrder(); // Can cancel in ORDER_PLACED state
    order2.nextState();  // Cannot transition from cancelled state
    
    return 0;
}

File: 04_BehaviouralDesignPatterns/05_StatePattern/Main.java
interface OrderState {
    void nextState(OrderContext context);

    void cancelOrder(OrderContext context);

    String getStateName();
}

class OrderPlacedState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is now being processed.");
        context.setState(new OrderProcessingState());
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order has been cancelled.");
        context.setState(new OrderCancelledState());
    }

    @Override
    public String getStateName() {
        return "Order Placed";
    }
}

class OrderProcessingState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is now being prepared.");
        context.setState(new OrderPreparedState());
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order cannot be cancelled while processing.");
    }

    @Override
    public String getStateName() {
        return "Order Processing";
    }
}

class OrderPreparedState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is now ready for delivery.");
        context.setState(new OrderReadyForDeliveryState());
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order cannot be cancelled after preparation.");
    }

    @Override
    public String getStateName() {
        return "Order Prepared";
    }
}

class OrderReadyForDeliveryState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is out for delivery.");
        context.setState(new OrderOutForDeliveryState());
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order cannot be cancelled after being prepared for delivery.");
    }

    @Override
    public String getStateName() {
        return "Order Ready for Delivery";
    }
}

class OrderOutForDeliveryState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order has been delivered.");
        context.setState(new OrderDeliveredState());
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order cannot be cancelled after being out for delivery.");
    }

    @Override
    public String getStateName() {
        return "Order Out for Delivery";
    }
}

class OrderDeliveredState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is already delivered. No further state transitions available.");
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order cannot be cancelled after delivery.");
    }

    @Override
    public String getStateName() {
        return "Order Delivered";
    }
}

class OrderCancelledState implements OrderState {
    @Override
    public void nextState(OrderContext context) {
        System.out.println("Order is cancelled. No further state transitions available.");
    }

    @Override
    public void cancelOrder(OrderContext context) {
        System.out.println("Order is already cancelled.");
    }

    @Override
    public String getStateName() {
        return "Order Cancelled";
    }
}

class OrderContext {
    private OrderState state;

    public OrderContext() {
        this.state = new OrderPlacedState();
    }

    public void setState(OrderState state) {
        this.state = state;
    }

    public void nextState() {
        state.nextState(this);
    }

    public void cancelOrder() {
        state.cancelOrder(this);
    }

    public String getStateName() {
        return state.getStateName();
    }
}

public class Main {
    public static void main(String[] args) {
        OrderContext order = new OrderContext();

        System.out.println("Current State: " + order.getStateName());
        order.nextState(); // Move to processing
        System.out.println("Current State: " + order.getStateName());

        order.cancelOrder(); // Attempt to cancel while processing
        System.out.println("Current State: " + order.getStateName());

        order.nextState(); // Move to prepared
        System.out.println("Current State: " + order.getStateName());

        order.nextState(); // Move to ready for delivery
        System.out.println("Current State: " + order.getStateName());

        order.nextState(); // Move to out for delivery
        System.out.println("Current State: " + order.getStateName());

        order.nextState(); // Move to delivered
        System.out.println("Current State: " + order.getStateName());

        order.cancelOrder(); // Attempt to cancel after delivery
        System.out.println("Current State: " + order.getStateName());
    }
}